renovate-ppc/src/Renovate/Arch/PPC/ABI.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE ScopedTypeVariables, TypeApplications, DataKinds, GADTs #-}
+
+-- | An 'ABI' implementation for the PPC ABI
+
+module Renovate.Arch.PPC.ABI ( abi64
+                             , abi32
+                             , instrOpcode
+                             ) where
+
+import qualified Data.Macaw.PPC as PPC
+import qualified Dismantle.PPC  as D
+import qualified Renovate       as R
+import           Renovate.Arch.PPC.ISA
+
+import qualified Data.Parameterized.Some as Some
+
+-- TODO: some of this could be moved into an Internals.hs
+
+-- | Note that allocateMemory, computeStackPointerOffset, saveReturnaddress, and
+-- checkShadowStack, are currently used only for the Shadow stack embrittle
+-- transforamtion, which we are not applying on PPC right now. As a result, we
+-- leave these undefined for the time being.
+abi64 :: R.ABI PPC.PPC64
+abi64 = R.ABI { R.isReturn            = ppcIsReturn . R.toGenericInstruction @PPC.PPC64 
+            , R.callerSaveRegisters = ppcCallerSaveRegisters64
+            , R.clearRegister       = fmap (const NoAddress) 
+                                    . R.fromGenericInstruction @PPC.PPC64 
+                                    . ppcClearRegister
+            , R.pointerSize         = 8
+            -- | these are all for shadow stack; leave undefined
+            , R.allocateMemory            = undefined
+            , R.computeStackPointerOffset = undefined
+            , R.saveReturnAddress         = undefined
+            , R.checkShadowStack          = undefined
+            }
+
+-- | Note that allocateMemory, computeStackPointerOffset, saveReturnaddress, and
+-- checkShadowStack, are currently used only for the Shadow stack embrittle
+-- transforamtion, which we are not applying on PPC right now. As a result, we
+-- leave these undefined for the time being.
+abi32 :: R.ABI PPC.PPC32
+abi32 = R.ABI { R.isReturn            = ppcIsReturn . R.toGenericInstruction @PPC.PPC32
+            , R.callerSaveRegisters = ppcCallerSaveRegisters32
+            , R.clearRegister       = fmap (const NoAddress) 
+                                    . R.fromGenericInstruction @PPC.PPC32
+                                    . ppcClearRegister
+            , R.pointerSize         = 4
+            -- | these are all for shadow stack; leave undefined
+            , R.allocateMemory            = undefined
+            , R.computeStackPointerOffset = undefined
+            , R.saveReturnAddress         = undefined
+            , R.checkShadowStack          = undefined
+            }
+
+
+-- | Retrieve the 'String' opcode from an instruction (for diagnostic purposes)
+instrOpcode :: D.Instruction -> String
+instrOpcode (D.Instruction opcode operands) = show opcode
+
+-- is return if it matches BLR (branch to link register) or BLRL 
+ppcIsReturn :: D.Instruction -> Bool
+ppcIsReturn instr = instrOpcode instr == "BLR" || instrOpcode instr == "BLRL"
+
```

Comment:
Could we change this to pattern matching on the actual BLR and BLRL constructors from Dismantle.PPC just so that we get a guarantee about the opcodes existing?